feat: PX-5157 - Add Bulk Edit endpoint

Mitch-Henson opened the pull request:
Solves [PX-5157]

This PR adds the Bulk Edit endpoint from Gravity into Metaphysics so that it can be consumed by coming changes in Volt.

As part of these changes, I've introduced some more beginner/backend friendly documentation on adding a REST endpoint into Metaphysics. As a first time user of GraphQL I definitely struggled even with the most basic of operations and hopefully this will help other future devs with their first PR.

[PX-5157]: https://artsyproduct.atlassian.net/browse/PX-5157?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

damassi commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+
+### Steps to add an existing endpoint
+
+1. Find an existing PR or commit which adds something similar to what you want. If you are looking for a PUT endpoint, then [this](https://github.com/artsy/metaphysics/commit/987b21f501bb3483d196720d4a252c25b704352e) is a great example.
```
It might be a bit better to link directly to a PR, versus a commit

Mitch-Henson replied:
Updated in the new branch. Will make a PR soon and link it back here

damassi commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+
+1. Find an existing PR or commit which adds something similar to what you want. If you are looking for a PUT endpoint, then [this](https://github.com/artsy/metaphysics/commit/987b21f501bb3483d196720d4a252c25b704352e) is a great example.
+1. Follow the structure of the PR to add your own endpoint and input/outputs. Some further tips can be found below.
+1. After the code has been written, you can run `yarn dump:local` in the Metaphysics directory to generate the new schema files. This script (and other useful scripts) can be found in the `package.json` file.
```
Why is this step needed?

Mitch-Henson replied:
I had problems with my local environment and the schema not being updated which caused errors until `yarn dump:local` was run. I can merge this into a single step with `yarn start` below with some additional information

damassi commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+1. After the code has been written, you can run `yarn dump:local` in the Metaphysics directory to generate the new schema files. This script (and other useful scripts) can be found in the `package.json` file.
+1. Start your local server with `yarn start`
+1. With your favourite graphql app, you should be able to hit the local server to refresh the schema. In insomnia, you can click on the "schema" button in the GraphQL panel to manually refresh this, if you don't have auto-refresh turned on (also found in this list)
+1. You should be able to hit your local metaphysics, which will hit staging gravity by default.
```
Having `local server` in the previous point and `local metaphysics` in this one is a bit confusing (since they actually mean the same thing). 

Maybe this point can just read "During development MP should be pointing at Gravity staging (unless otherwise configured in `.env.shared`)

Mitch-Henson replied:
Replaced all instances with `local Metaphysics server` and also updated the point you mentioned!

damassi commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+
+### Tips on writing and debugging locally
+
+- GraphQL is strongly typed and needs to know the exact inputs and outputs of the endpoint as it mediates between the clients and the REST endpoint. This is why you need to set up a union type which instructs GraphQL on what both responses will look like. This allows GraphQL tools to know if the request you are making (in Insomnia, for example) is valid, even before you send the request.
```
Going from setup to writing mutation-handling Union types is a very steep curve! 

As we want this document to be instructive (rather than intimidating) its best to keep this all very incremental versus skipping ahead. Meaning: 

- Add an example that shows how to add a new loader pointing at a Gravity `GET` endpoint 
- Show how to add a new field to a section within the schema (say, `artwork.ts`) 
- Show how to add a new `POST` endpoint in prep for a mutation
- Show had to add a new mutation (in the simplest way possible; no fancy error handling is necessary yet; see all of the mutations in the `me` folder for an example)
- THEN show how to handle errors in a way that is more sophisticated. 

But even so, this `thingOrError` pattern isn't necessarily something we should advocate for up front in a getting started doc. In many ways it's overly complicated for everyday mutations and should be used more on a need to know basis. 

(Worth noting that we have a full blog post on this here: https://artsy.github.io/blog/2018/10/19/where-art-thou-my-error/)

Mitch-Henson replied:
Thanks for the article link, I will have to check it out and also link it in this doc.

I guess I'm simultaneously both the best and worst person to make this doc. The best person because I know almost nothing about GraphQL and can give feedback on whether something is too complicated or not, and also the worst person because I know nothing about GraphQL and so I shouldn't be teaching others about things I don't really understand.

The `thingOrError` I thought was standard and came from the [example PR](https://github.com/artsy/metaphysics/pull/3918) I was basing my changes on. Happy to learn the easier way on how to do this!

damassi commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
````diff
@@ -0,0 +1,78 @@
+  },
+```
+
+- You don't need to restart the server once you make changes. The updates get triggered automatically once you save your changes locally and hit your local metaphysics server again. You should see some lines similar to these below, which shows that it has been updated
````
> hit your local metaphysics server again

to 

> re-run your query in GraphiQL (or whatever app you use)

Mitch-Henson replied:
Updated!

mzikherman commented on src/schema/v2/partnerArtworksMutation.ts:
```diff
@@ -0,0 +1,136 @@
+    { id, artsyShippingDomestic, artsyShippingInternational, location },
+    { updatePartnerArtworksLoader }
+  ) => {
+    if (!updatePartnerArtworksMutation) {
```
Not sure when `!updatePartnerArtworksMutation` would be true. Did you mean to check if the loader exists (which would indicate the MP request was lacking an `X-Access-Token` header if it didn't)?

Mitch-Henson replied:
I was following [this PR](https://github.com/artsy/metaphysics/pull/3918) as this REST endpoint is only available for admins at the moment, and this seemed to match.

Mitch-Henson replied:
I've removed this check as the admin only feature is actually already on the REST endpoint so this is redundant.

mzikherman replied:
If you notice that PR checks the _loader_ existence (not the mutation). It's still good to check the loader existence in order to fail with a valid error ('This operation only available with an access token'), vs. now where this mutation might just return `null` (w/ no error or explanation) if a user attempts it w/o a token.

MP doesn't verify tokens, but it does use the existence of them being passed as headers to decide if the authenticated loaders (and typically most mutations) have loaders made available in the context.

Mitch-Henson replied:
Totally missed that, it all makes sense now. Thanks for the context! Just added the checking of the loader in, so now the null check is no longer needed

mzikherman commented on src/schema/v2/partnerArtworksMutation.ts:
```diff
@@ -0,0 +1,136 @@
+      artsy_shipping_international: artsyShippingInternational,
+      location,
+    }
+    return updatePartnerArtworksLoader?.(id, gravityOptions)
```
By checking the loader above, can get rid of the `?`. Also, it's preferable to use `async` / `await` instead of `.then` structure.

Mitch-Henson replied:
Updated to use `async` / `await`

Mitch-Henson replied:
I can't seem to figure out why this can potentially be `null`. Using the [createAccountRequestMutation](https://github.com/artsy/metaphysics/blob/d7bdf8300427c70aec30b8c55737770e5ebcb277/src/schema/v2/createAccountRequestMutation.ts#L88-L124) as an example, I can't seem to find the difference.

mzikherman replied:
Ah, commented above, but:

> MP doesn't verify tokens, but it does use the existence of them being passed as headers to decide if the authenticated loaders (and typically most mutations) have loaders made available in the context.

So someone issuing this mutation without providing an `X-Access-Token` header, this loader would be `null`. That happens in: https://github.com/artsy/metaphysics/blob/d7bdf8300427c70aec30b8c55737770e5ebcb277/src/lib/loaders/index.ts#L82-L83

Mitch-Henson replied:
It all makes sense now. Thanks!

mzikherman commented on src/schema/v2/partnerArtworksMutation.ts:
```diff
@@ -0,0 +1,136 @@
+  ResolverContext
+>({
+  name: "UpdatePartnerArtworksMutationSuccess",
+  isTypeOf: (data) => data.success || data.errors,
```
Seems confusing, a `Success` type can be returned if the underlying data contains `errors`?

Mitch-Henson replied:
I guess it can be confusing without some context. The REST endpoint response looks like so

```javascript
// All artworks updated
{
  "success": 4,
  "errors": {
    "count": 0,
    "ids": []
  }
}

// Some / no artworks updated
{
  "success": 0,
  "errors": {
    "count": 4,
    "ids": [
      "5ab3ea0b139b211428d47600",
      "5c7ecd5ecb5b32002879ff59",
      "5c7ecd5e53241d002c6b30c0",
      "5c7ecd5ef5464e269088ff9f"
    ]
  }
}
```

What Leandro and I found is that if we just put `isTypeOf: (data) => data.success` then it fails if `success` is zero. The `errors` will always be valid, even if empty, because it is an object which is inherently true, as opposed to `0` which can be interpreted as `false`. The endpoint successfully returns, but we can't know how many artworks (if any) will be updated ahead of time, due to other factors.

mzikherman replied:
Ah yea, if it's an integer and can be 0 as 'success', there's an `isExisty` helper (https://github.com/artsy/metaphysics/blob/d7bdf8300427c70aec30b8c55737770e5ebcb277/src/lib/helpers.ts#L33) you could use, probably `isExisty(data.success)`.

Mitch-Henson replied:
Perfect, updated!

mzikherman commented on src/schema/v2/partnerArtworksMutation.ts:
```diff
@@ -0,0 +1,136 @@
+  location: string | null
+}
+
+const PartnerArtworksType = new GraphQLObjectType<any, ResolverContext>({
```
This type might be better named as `UpdatePartnerArtworksMutationSuccessDetails` or something? (caveat is that's verbose and I'm not a great namer 😆 )

Mitch-Henson replied:
You are totally right! I've updated this as the name is verbose but explains exactly what it is. Also fits with the existing types.

damassi reviewed (commented):
(moved comment)

mzikherman reviewed (dismissed):
LGTM and fine to merge/iterate. I think the `PartnerArtworks` name should be improved, and I'm a bit confused as to the underlying bulk endpoint (seems like a successful mutation might still include errors (like for individual artwork errors?)) so it's a _bit_ of an odd duck but totally fine to iterate on!

damassi reviewed (changes requested):
@Mitch-Henson - the docs are a good start! I would suggest pulling them out of this PR and into something separate which we can develop outside of feature work and get some more reviewers.  

(But yeah, looking [here](https://github.com/artsy/metaphysics/blob/main/docs/graphql-schema-design.md) I agree that our docs are woefully inadequate for introducing a new dev to GraphQL. I think if we start slow and then build we could get a simple guide going for folks that would reduce the learning curve substantially.)

Mitch-Henson commented on _schemaV2.graphql:
```diff
@@ -15450,6 +15471,38 @@ type updateNotificationPreferencesMutationPayload {
+
+  # Whether Artsy international shipping should be enabled
+  artsyShippingInternational: Boolean
+  clientMutationId: String
```
I'm assuming this is coming from using `mutationWithClientMutationId`? I had a quick cursory scan and it seems like all other `*Mutation.ts` files in the same directory use this approach. Absolutely for changing this, just not sure what to change it to. I see that the `InternalIDFields` adds the `clientMutationId` to the response schema, but not to the request

mzikherman replied:
Yea, that's for some compatibility with really old versions of Relay (which we don't use). Most likely it's not needed (and we should maybe not need/use that specific helper anymore), but doesn't have to be updated in this PR!

Mitch-Henson commented on _schemaV2.graphql:
```diff
@@ -11705,6 +11710,22 @@ type PartnerArtworkGrid implements ArtworkContextGrid {
   title: String
 }
 
+type PartnerArtworks {
```
Updated with inspiration from [your comment here](https://github.com/artsy/metaphysics/pull/4269#discussion_r934700874)

Mitch-Henson commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+1. Start your local server with `yarn start`
+1. With your favourite graphql app, you should be able to hit the local server to refresh the schema. In insomnia, you can click on the "schema" button in the GraphQL panel to manually refresh this, if you don't have auto-refresh turned on (also found in this list)
+1. You should be able to hit your local metaphysics, which will hit staging gravity by default.
+1. Work through the errors, making sure to read the message in its entirety. Also don't hesitate to add it to the errors section below if useful.
```
I based this more or less off of my own experience, so it can definitely be improved for someone else.

With this section I was trying to prepare the developer to see errors (you are right in that this should be specified for what errors and where) but also to encourage further additions to the documentation if someone comes across an error that could be documented to help someone else get setup quicker.

I don't mind if we remove this point from these steps, however I believe there definitely needs to be a section on debugging and error expectations

Mitch-Henson commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+    partnerArtworksOrError: {
+      type: UpdatePartnerArtworksMutationType,
+      resolve: (result) => {
+        console.log(result)
```
What are you thoughts about leaving the console log in because it is very basic, but also including a [link to the debugger approach](https://github.com/artsy/metaphysics/blob/main/docs/debugging_with_vscode.md)?

damassi replied:
That sounds good 👍 (And I didn't know about this doc!)

Mitch-Henson commented on docs/adding_a_new_gravity_endpoint_into_metaphysics.md:
```diff
@@ -0,0 +1,78 @@
+
+### Potential Errors
+
+This below error points to the fact that the `UpdatePartnerArtworksMutationType` was not resolving for either the `SuccessType` or the `FailureType`. GraphQL is very strict and it needs to categorize everything into types and fails when it can't. The solution to this was to ensure that the `SuccessType` was correctly defined. This means that the `isTypeOf` check needs to look for a success criter ...
```
This is based on my personal experience and I wanted to have a potential errors section where we can list generic errors that might help people find the root cause of their own errors, without relying on someone with more GraphQL knowledge. I didn't pay attention to all the errors that I experienced before I decided to write this doc so unfortunately I only have this one. I found it super helpful after understanding the error on how the code determines whether something is a success or an error. I think it is useful here (something along these lines, of the response not being a success or an error) however happy to move it somewhere else talking generally about the job of each function

Mitch-Henson commented:
Thanks for the feedback so far!

As per @damassi suggestion, I've moved the docs into its own PR here: https://github.com/artsy/metaphysics/pull/4273 so that it doesn't block feature work.

@mzikherman I've addressed most of your concerns except I'm still struggling with removing the [clientMutationId](https://github.com/artsy/metaphysics/pull/4269#discussion_r934695227) from the input and also making sure that the `updatePartnerArtworksLoader` is [never null](https://github.com/artsy/metaphysics/pull/4269#discussion_r934698501).

mzikherman commented on _schemaV2.graphql:
```diff
@@ -11712,6 +11717,11 @@ type PartnerArtworkGrid implements ArtworkContextGrid {
   title: String
 }
 
+type PartnerArtworksError {
```
similar comment as `PartnerArtworks` naming, should this be `PartnerArtworksBulkEditErrors` or something?

Mitch-Henson replied:
Updated to that, thanks!

mzikherman commented on _schemaV2.graphql:
```diff
@@ -15458,6 +15468,49 @@ type updateNotificationPreferencesMutationPayload {
+}
+
+type UpdatePartnerArtworksMutationSuccess {
+  partnerArtworks: UpdatePartnerArtworksMutationSuccessDetails
```
Naming here as well, the field name might be better off as something besides `partnerArtworks`.

Mitch-Henson replied:
Updated to `partnerArtworksBulkEdit`

mzikherman commented on _schemaV2.graphql:
```diff
@@ -15458,6 +15468,49 @@ type updateNotificationPreferencesMutationPayload {
+  id: ID!
+
+  # A type-specific ID likely used as a database ID.
+  internalID: ID!
```
Is there a database ID returned from that PUT? (doesn't look like it based on https://github.com/artsy/gravity/blob/53acd88de5deb8b8c3ac7c84cc57cc0a31660618/spec/api/v1/partners_artworks_endpoint_spec.rb#L759)

mzikherman commented on src/schema/v2/partnerArtworksMutation.ts:
```diff
@@ -0,0 +1,134 @@
+>({
+  name: "UpdatePartnerArtworksMutationSuccessDetails",
+  fields: () => ({
+    ...InternalIDFields,
```
probably doesn't need this

Mitch-Henson replied:
Removed!

damassi reviewed (commented):
Thanks for moving the docs into its own PR 👍

mzikherman commented:
Looks great, thanks for bearing with all the comments! Could use a spec but otherwise pretty perfect.

Mitch-Henson commented:
Whoops, you can ignore the re-review request @damassi. It hadn't loaded correctly and looked like your previous review was blocking, even though you dismissed it.

I will merge this in now to unblock some Volt work, but will add some tests and tag you in that follow-up PR @mzikherman
